lighthouse-core/audits/byte-efficiency/unused-javascript.js

```diff
@@ -0,0 +1,184 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./byte-efficiency-audit');
+const URL = require('../../lib/url-shim');
+
+const IGNORE_THRESHOLD_IN_BYTES = 2048;
+
+class UnusedJavaScript extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'unused-javascript',
+      description: 'Unused JavaScript',
+      informative: true,
+      helpText: 'Remove unused JavaScript to reduce bytes consumed by network activity.',
+      requiredArtifacts: ['JsUsage', 'networkRecords']
+    };
+  }
+
+  /**
+   * @param {{used: number, unused: number}} objA
+   * @param {{used: number, unused: number}} objB
+   * @return {{used: number, unused: number}}
+   */
+  static _add(objA, objB) {
+    return {
+      used: objA.used + objB.used,
+      unused: objA.unused + objB.unused,
+    };
+  }
+
+  /**
+   * Determines the length of unused and used portions in a set of ranges.
+   * Because the protocol reports usage for each function and ranges within each function, nested
+   * regions are represented numerous times. This function makes sure to only count each string
+   * section once. See unit and smoke tests for examples.
+   *
+   * @param {!Array<{isUsed: boolean, startOffset: number, endOffset: number
+   *    used: number, unused: number}>} ranges
+   * @return {{used: number, unused: number}}
+   */
+  static findUsedAndUnused(ranges) {
+    if (ranges.length === 0) {
+      return {used: 0, unused: 0};
+    } else if (ranges.length === 1) {
+      return ranges[0];
+    }
+
+    const firstRange = ranges[0];
+    const secondRange = ranges[1];
+    if (secondRange.startOffset >= firstRange.endOffset) {
+      // Ranges aren't nested, we can simply continue on to the rest.
+      return this._add(
+        this._add(firstRange, {used: secondRange.startOffset - firstRange.endOffset, unused: 0}),
```

Comment:
that this should be considered `used` isn't immediately obvious. Is it non-functional code (from my speculation above) and so we want to consider it wasted? A comment would be good that we're creating our own range to cover the gap and why it's treated as used